generator/abstractmetabuilder.cpp

```diff
@@ -2480,55 +2480,100 @@ void AbstractMetaBuilder::dumpLog()
 
 AbstractMetaClassList AbstractMetaBuilder::classesTopologicalSorted() const
 {
-    AbstractMetaClassList res;
-
-    AbstractMetaClassList classes = m_meta_classes;
-    classes.sort();
-
-    QSet<AbstractMetaClass*> noDependency;
-    QHash<AbstractMetaClass*, QSet<AbstractMetaClass* >* > hash;
-    for (AbstractMetaClass *cls :  classes) {
-        QSet<AbstractMetaClass* > *depends = new QSet<AbstractMetaClass* >();
-
+    /* This function is the standard topological of a Directed Acyclic Graph
+     * (a DAG).  It outputs a partially ordered list of the nodes in the graph
+     * such that a node is output after all of its children.
+     *
+     * In the previous implementation it accounted for around 68-69% of the
+     * entire run time of pythonqt_generator.  The implementation also leaked
+     * memory and modified a QHash while iterating over it with a QHashIterator.
+     *
+     * This version is somewhat like the python equivalent using list
+     * comprehensions.
+     */
+
+    /* Build a hash list of QSet<class> for each class.  'class' is represented
+     * by a pointer to the AbstractMetaClass from m_meta_classes.
+     */
+    ReportHandler::debugSparse(QString("TSORT: %1 meta classes")
+            .arg(m_meta_classes.count()));
+    QHash<AbstractMetaClass*, QSet<AbstractMetaClass*>> classes;
+    classes.reserve(m_meta_classes.count());
+
+    for (auto cls : m_meta_classes) {
+        /* Add the baseClass and the interfaces the class uses.  The latter
+         * are stored in an AbstractMetaClassList which is uses a QList, so:
+         */
+        auto entry(classes.insert(cls, QSet<AbstractMetaClass*>(
+                    cls->interfaces().cbegin(), cls->interfaces().cend())));
```

Comment:
This is not upwards compatible. Using 
```cpp
#if QT_VERSION >= 0x060000
        auto interfaces = QSet<AbstractMetaClass*>(
            cls->interfaces().cbegin(), cls->interfaces().cend());
#else
        auto interfaces = QSet<AbstractMetaClass*>::fromList(cls->interfaces());
#endif
        auto entry(classes.insert(cls, interfaces));
```
instead fixes it. Actually `cBegin()` / `cEnd()` have been introduced somewhere in Qt 5.15, so the check could  be changed, but I just did the same as in other places in the code.